internal/core/metadata/operators/device/db.go

```diff
@@ -32,3 +29,23 @@ type DeviceUpdater interface {
 	DeviceProfileLoader
 	DeviceServiceLoader
 }
+
+type DeviceProfileUpdater interface {
+	GetAllDeviceProfiles() ([]contract.DeviceProfile, error)
+	GetProvisionWatchersByProfileId(id string) ([]contract.ProvisionWatcher, error)
+	UpdateDeviceProfile(dp contract.DeviceProfile) error
+	DeviceProfileLoader
+	DeviceLoader
+}
+
+type DeviceProfileLoader interface {
+	GetDeviceProfileById(id string) (contract.DeviceProfile, error)
+	GetDeviceProfileByName(n string) (contract.DeviceProfile, error)
+}
+
+type DeviceProfileDeleter interface {
+	DeleteDeviceProfileById(id string) error
+	DeviceProfileLoader
+	DeviceLoader
+	provsion_watcher.ProvisionWatcherLoader
+}
```

Comment:
Should these DeviceProfile-related interfaces be defined instead in internal/core/metadata/operators/device_profile/db.go?

Reply:
I moved them around due to cyclic dependencies.  The device-profile related interfaces require some of the device related interfaces. Is there a better way to organize these interfaces?

Reply:
To be more clear, if I move the device-profile related interfaces I get the following cycle:
```
DeviceAdder -> DeviceProfileLoader -> DeviceLoader
```

This is one of the few cycles that will exist as a result of the move.

However, the `DeviceProfileLoader` interface is not dependant on any of the device interfaces only the `device` interfaces depend on the `DeviceProfileLoader` which we should be able to move. Sadly, this also causes a cycle with the tests as the tests in `delete_test.go` and `update_test.go` rely on the interfaces in the `device` package.

Reply:
Got it. Remember though that Go uses interfaces differently from other languages. So we can define interfaces scoped to a given package. They don't need to be centrally defined and the emphasis on re-use of a single, shared interface definition is lessened. I personally don't think there's anything wrong with redeclaring interfaces that are scoped to a different package. In fact, I think mixing the approaches toward interfaces will get us into trouble w/r/t circular deps. See the following branch
https://github.com/tsconn23/edgex-go/tree/pr_1523_tsc